drivers/media/platform/raspberrypi/pisp_be/pisp_be.c

```diff
@@ -578,119 +538,103 @@ static int pispbe_schedule_internal(struct pispbe_node_group *node_group,
 			 * global enables aren't set for these blocks. If a
 			 * buffer has been provided, we dequeue it back to the
 			 * user with the other in-use buffers.
-			 *
 			 */
 			ignore_buffers = true;
 		}
 
 		node = &node_group->node[i];
 
-		spin_lock_irqsave(&node->ready_lock, flags1);
+		/* Pull a buffer from each V4L2 queue to form the queued job */
+		spin_lock_irqsave(&node->ready_lock, flags);
 		buf[i] = list_first_entry_or_null(&node->ready_queue,
 						  struct pispbe_buffer,
 						  ready_list);
-		spin_unlock_irqrestore(&node->ready_lock, flags1);
-		if (!buf[i] && !ignore_buffers) {
-			dev_dbg(pispbe->dev, "Nothing to do\n");
-			return 0;
-		}
-	}
+		if (buf[i])
+			list_del(&buf[i]->ready_list);
+		spin_unlock_irqrestore(&node->ready_lock, flags);
 
-	/* Pull a buffer from each V4L2 queue to form the queued job */
-	for (i = 0; i < PISPBE_NUM_NODES; i++) {
-		if (buf[i]) {
-			node = &node_group->node[i];
+		if (!buf[i] && !ignore_buffers)
+			return -ENODEV;
```

Comment:
Have we changed the logic here, by amalgamating these two loops, and is this safe to do/

Reply:
ok, do you mean

```
		if (!buf[i] && !ignore_buffers)
			return -ENODEV;

		pispbe->queued_job.buf[i] = buf[i];
```

we shouldn't `pispbe->queued_job.buf[i] = buf[i];` if `!buf[i]` ?

I thought setting it to (a possible) `NULL` was safe as that was its default state anyway. I can protect this with

```
 if (buf[i])
		pispbe->queued_job.buf[i] = buf[i];
```

or drop this patch completely, as even if there's a tiny window between the two loops, it should be safe ?

Reply:
I was concerned with the `list_del()` which looks like it could leak buffers in case of failure?

Normally the user will enqueue all the image buffers _before_ the config meta buffer, so this won't be reached. However, we don't want to allow a user error to cause a kernel error.